### ISSUE TITLE
[ms-compatibility] Recover better from missing 'typename' in sizeof

### ISSUE THREAD

**Issue by ec04fc15-fa35-46f2-80e1-5d271f2ef708** [llvm/llvm-project#18817]
|  |  |
| --- | --- |
| Bugzilla Link | [18443](https://llvm.org/bz18443) |
| Version | trunk |
| OS | Windows XP |
| CC | @majnemer,@rnk |

## Extended Description 
Consider:

struct X { typedef int type; };
template<typename T> void f() { sizeof(T::type); }
template void f<X>();

Clang diagnoses this as:

<stdin>:2:43: error: unexpected type name 'type': expected expression
template<typename T> void f() { sizeof(T::type); }
                                          ^
<stdin>:3:15: note: in instantiation of function template specialization 'f<X>' requested here
template void f<X>();
              ^

... but this sort of thing apparently happens inside MS system headers. We should support this case of missing 'typename' more elegantly -- perhaps we can teach TreeTransform to cope with this during instantiation.

**Comment by AaronBallman** [comment]
We currently issue this warning:
```
<source>:2:40: error: missing 'typename' prior to dependent type name 'X::type'
template<typename T> void f() { sizeof(T::type); }
                                       ^~~~~~~
                                       typename 
```
so I'm closing this as being implemented.
